### PR TITLE
Time to angle calculator

### PR DESCRIPTION
**Pull request by alexander-aleksandrov**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+language: java
+script: gradle check
```
Comment (alexander-yevsyukov) [inline .travis.yml]: Linefeed @ EOF. Please check here and in the rest of all the files. It's a usual problem. I gave you the link to how to fix it when saving files from IDEA.

File: src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/Main.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package com.teamdev.javaclasses.aleksandrov.brainfuck;
```
Comment (alexander-yevsyukov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/Main.java]: Please have copyright headers in all the files.

File: build.gradle [hunk 3]
```diff
@@ -20,6 +22,7 @@ ext.log4jVersion         = '2.6.2'
 apply plugin: 'java'
 apply plugin: 'eclipse'
 apply plugin: 'idea'
+apply plugin: 'groovy'
 apply plugin: 'org.junit.platform.gradle.plugin'
 
 jar {
```
Comment (alexander-yevsyukov) [inline build.gradle]: Please verify the content of this section. In particular, `jar.baseName` doesn't look right. It should reflect the name of the project. We don't really need packing it all into JAR.

Comment (alexander-yevsyukov) [inline build.gradle]: You also don't have slow or fast tags for your tests at this stage.

File: src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/Main.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+import com.teamdev.javaclasses.aleksandrov.brainfuck.compiler.BrainFuckToJavaCompilerImpl;
+import com.teamdev.javaclasses.aleksandrov.brainfuck.interpreter.BrainFuckInterpreterImpl;
+
+import java.io.File;
+
+public class Main {
+    static private final File output = new File ("C:/Projects/BrainFuck1/src/main/java/com/teamdev/javaclasses/files/BrainFuckRunner.java");
```
Comment (alexander-yevsyukov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/Main.java]: Do not use absolute paths like this. Problems:
1. They are Windows-only.
2. Even under Windows not all people would have them.

You can work with temporary directories, or you can require directories passed as program arguments.

File: src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/Main.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package com.teamdev.javaclasses.aleksandrov.brainfuck;
+
+import com.teamdev.javaclasses.aleksandrov.brainfuck.compiler.BrainFuckToJavaCompilerImpl;
+import com.teamdev.javaclasses.aleksandrov.brainfuck.interpreter.BrainFuckInterpreterImpl;
+
+import java.io.File;
+
+public class Main {
```
Comment (alexander-yevsyukov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/Main.java]: You can use `{@inheritDoc}` if there's nothing meaningful to add.

File: src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/comand/Memory.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+    private final int[] cells;
+    private int pointer = 0;
+
+    public Memory(int memorySize) {
+
+        cells = new int[memorySize];
+    }
+
+    public void setCurrentCellValue(final int newValue) {
```
Comment (alexander-yevsyukov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/comand/Memory.java]: Do not use `final` unless an argument is used in a `Runnable` or another multi-threading code, which _requires_ the argument to be `final`.

This rule helps finding multi-threading.

File: src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/compiler/BrainFuckToJavaCompilerVisitor.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.List;
+
+public class BrainFuckToJavaCompilerVisitor implements CommandVisitor {
+    private final File output = new File ("C:/Projects/BrainFuck1/src/main/java/com/teamdev/javaclasses/files/Draft.txt");
+    private final Printer printer = new FilePrinterImpl(output);
+
+
+    @Override
+    public void visit(CellDecrement cellDecrement) {
+        List<String> text = new ArrayList<>();
```
Comment (alexander-yevsyukov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/compiler/BrainFuckToJavaCompilerVisitor.java]: Please use `Lists.newArrayList()` from Guava.

File: src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/files/Draft.txt [hunk 8]
```diff
@@ -0,0 +1,2 @@
+compiledChar = (char) memory[pointer];
```
Comment (alexander-yevsyukov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/files/Draft.txt]: Are these test files? They should not belong to `java` directory. By convention such things below to `resources`.

File: src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/Main.java (hunk 2 above)
Comment (alexander-yevsyukov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/Main.java]: Have `package-info.java` with appropriate annotations that treat nullity.

File: src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/reader/FileReaderImp.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+        try {
+            Scanner in  = new Scanner(new FileReader(file));
+            return in.next();
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
+        }
+        return "";
+    }
+    }
+
```
Comment (alexander-yevsyukov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/reader/FileReaderImp.java]: Please fix the file layout and empty space here.

File: src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/EvaluationStack.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
+package com.teamdev.javaclasses.aleksandrov.calculator;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class EvaluationStack {
+
+    private final static Logger LOG = Logger.getLogger(EvaluationStack.class.getName());
```
Comment (alexander-yevsyukov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/EvaluationStack.java]: We normally use [lazy logging](https://sites.google.com/a/teamdev.com/engineers/development-standards/java/style-guide/logging#TOC-Lazily-initialized), which is based on Slf4J. Please migrate.

File: src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/ParsingService.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.logging.Logger;
+
+public class ParsingService implements StateRecognizer<
+        ExpressionReader,
+        EvaluationStack,
+        CalculationState> {
+
+    private final static Logger LOG = Logger.getLogger(StateRecognizer.class.getName());
+
+    final private Map<CalculationState, ExpressionParser> parsers = new HashMap<CalculationState, ExpressionParser>() {{
```
Comment (alexander-yevsyukov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/ParsingService.java]: Don't use such “inheritance”. Use `ImmutableMap.Builder` from Guava or other builders if you need mutable instances.

File: src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/ParsingService.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+        put(CalculationState.NUMBER, new NumberParser());
+        put(CalculationState.BINARY_OPERATOR, new BinaryOperatorParser());
+        put(CalculationState.FINISH, new EndOfExpressionParser());
+    }};
+
+    public boolean accept(ExpressionReader reader, EvaluationStack stack, CalculationState possibleState) {
+        final ExpressionParser parser = parsers.get(possibleState);
+
+        if (LOG.isLoggable(Level.INFO)) {
```
Comment (alexander-yevsyukov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/ParsingService.java]: Use Slf4J and `{}`-based formatting. This way you won't need to check for the logging level and do not have performance impact if the logging is turned on.

File: src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/ParsingService.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+            LOG.info("Next posible state is: " + possibleState);
+        }
+
+        if (parser == null) {
+            throw new IllegalStateException("ParserImpl not found for state: " + possibleState);
+        }
+        if (LOG.isLoggable(Level.INFO)) {
```
Comment (alexander-yevsyukov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/ParsingService.java]: Once again, if a logging level is turned off, and you don't have parameters, there's little sense to check for the log level status. Use Slf4J.

File: src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/finitestatemachine/MachineRunner.java [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+            if (currentState == null) {
+                deadlock(inputContext);
+            }
+        }
+
+        return prepareResult(outputContext);
+    }
+
+    private State acceptNewState(InputContext inputContext, OutputContext outputContext,
```
Comment (alexander-yevsyukov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/finitestatemachine/MachineRunner.java]: Align parameters vertically if they don't fit into a single line. Adjust IDEA accordingly.

File: src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/finitestatemachine/MachineRunner.java [hunk 15]
```diff
@@ -0,0 +1,75 @@
+package com.teamdev.javaclasses.aleksandrov.calculator.finitestatemachine;
+
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+
+abstract public class MachineRunner<
```
Comment (alexander-yevsyukov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/finitestatemachine/MachineRunner.java]: I think that a canonical order would be `public abstract`. There's IDEA inspection for this. Take default inspection values from the [Engineers site](https://sites.google.com/a/teamdev.com/engineers/development-standards/idea-configuration).

File: src/main/java/com/teamdev/javaclasses/aleksandrov/phonenumber/Country.java [hunk 16]
```diff
@@ -0,0 +1,14 @@
+package com.teamdev.javaclasses.aleksandrov.phonenumber;
+
+public enum Country {
+    UNITED_STATES("United States"),
+    CANADA("Canada"),
+    NEWCOUNTRY("qwerty");
+
+    private String s;
```
Comment (alexander-yevsyukov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/phonenumber/Country.java]: `s` is a really bad name for a field.

File: src/main/java/com/teamdev/javaclasses/aleksandrov/phonenumber/PhoneNumber.java [hunk 17]
```diff
@@ -0,0 +1,116 @@
+package com.teamdev.javaclasses.aleksandrov.phonenumber;
+
+import com.sun.istack.internal.Nullable;
+
+import java.text.ParseException;
+
+public final class PhoneNumber {
+
+    @Nullable
+
```
Comment (alexander-yevsyukov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/phonenumber/PhoneNumber.java]: Why empty space here?

File: src/main/java/com/teamdev/javaclasses/aleksandrov/phonenumber/PhoneNumber.java [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+        input.toCharArray();
+
+
+    }
+
+
+    @Override
+    public int hashCode() {
+        return 0;
```
Comment (alexander-yevsyukov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/phonenumber/PhoneNumber.java]: NO.

File: src/main/java/com/teamdev/javaclasses/aleksandrov/phonenumber/PhoneNumber.java [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return true;
```
Comment (alexander-yevsyukov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/phonenumber/PhoneNumber.java]: NO.

**Comment by alexander-yevsyukov** [comment]
And, BTW, the PR name does not reflect the content. Please adjust.

File: src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/Main.java (hunk 2 above)
Comment (alexander-aleksandrov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/Main.java]: Fixed 
JFYI: Oleg said that this is a garbage in a code. That is why I always delete them.

File: .travis.yml (hunk 1 above)
Comment (alexander-aleksandrov) [inline .travis.yml]: Fixed

File: build.gradle (hunk 3 above)
Comment (alexander-aleksandrov) [inline build.gradle]: Fixed

File: src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/comand/Memory.java (hunk 6 above)
Comment (alexander-aleksandrov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/comand/Memory.java]: Fixed

File: src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/comand/NextCell.java [hunk 20]
```diff
@@ -0,0 +1,11 @@
+package com.teamdev.javaclasses.aleksandrov.brainfuck.comand;
+
+import com.teamdev.javaclasses.aleksandrov.brainfuck.compiler.CommandVisitor;
+
+public class NextCell implements Command{
```
Comment (alexander-aleksandrov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/comand/NextCell.java]: Fixed

File: src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/compiler/BrainFuckToJavaCompilerImpl.java [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+        final Reader reader = new FileReaderImp();
+        final Parser parser = new InterpreterParser();
+        final Generator generator = new JavaFileGenerator();
+        final JavaFileInserter javaFile = new JavaFileInserter();
+        final String programText;
+
+        programText = reader.read(program);
+        generator.generateTemplate(output);
+
+
```
Comment (alexander-aleksandrov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/compiler/BrainFuckToJavaCompilerImpl.java]: Fixed

File: src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/compiler/BrainFuckToJavaCompilerImpl.java [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+        final BrainFuckToJavaCompilerVisitor visitor =
+                new BrainFuckToJavaCompilerVisitor();
+
+        for (Command command : commands) {
+            command.accept(visitor);
+        }
+
+        javaFile.generateJavaFile();
+
```
Comment (alexander-aleksandrov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/compiler/BrainFuckToJavaCompilerImpl.java]: Fixed

File: src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/files/Draft.txt (hunk 8 above)
Comment (alexander-aleksandrov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/files/Draft.txt]: Fixed

File: src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/reader/FileReaderImp.java (hunk 9 above)
Comment (alexander-aleksandrov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/reader/FileReaderImp.java]: Fixed

File: src/main/java/com/teamdev/javaclasses/aleksandrov/phonenumber/PhoneNumber.java (hunk 17 above)
Comment (alexander-aleksandrov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/phonenumber/PhoneNumber.java]: Fixed

File: src/main/java/com/teamdev/javaclasses/aleksandrov/phonenumber/Country.java (hunk 16 above)
Comment (alexander-aleksandrov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/phonenumber/Country.java]: Fixed

File: src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/finitestatemachine/MachineRunner.java (hunk 14 above)
Comment (alexander-aleksandrov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/finitestatemachine/MachineRunner.java]: Fixed

File: src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/Main.java (hunk 2 above)
Comment (alexander-aleksandrov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/Main.java]: Fixed

File: src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/finitestatemachine/MachineRunner.java (hunk 15 above)
Comment (alexander-aleksandrov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/finitestatemachine/MachineRunner.java]: Fixed

File: src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/Main.java (hunk 2 above)
Comment (alexander-yevsyukov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/Main.java]: Copyright headers are needed for understanding the license conditions under which the code is issued. It is also a legal protection for the company who release the code.

So, no, this is not a “garbage”. It is collapsed automatically by IDEA (as well as updated or inserted. Find a command for it.) So it does not hinder a developer by no means.

Review the code of Google. They insert such comments even in build scripts.

File: src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/ParsingService.java (hunk 13 above)
Comment (alexander-aleksandrov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/ParsingService.java]: Fixed

File: src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/EvaluationStack.java (hunk 10 above)
Comment (alexander-aleksandrov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/EvaluationStack.java]: Fixed

File: src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/compiler/BrainFuckToJavaCompilerVisitor.java (hunk 7 above)
Comment (alexander-aleksandrov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/compiler/BrainFuckToJavaCompilerVisitor.java]: Fixed

File: src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/ParsingService.java (hunk 12 above)
Comment (alexander-aleksandrov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/ParsingService.java]: Fixed

File: src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/ParsingService.java (hunk 11 above)
Comment (alexander-aleksandrov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/calculator/ParsingService.java]: Fixed

File: build.gradle [hunk 23]
```diff
@@ -1,73 +1,47 @@
...
-    baseName = 'junit5-gradle-consumer'
-    version = '1.0.0-SNAPSHOT'
-}
 
-compileTestJava {
-    sourceCompatibility = 1.8
-    targetCompatibility = 1.8
-    options.compilerArgs += '-parameters'
+repositories {
+      mavenCentral()
```
Comment (alexander-yevsyukov) [inline build.gradle]: You may also encounter`jcenter()` in build scripts. They say it's faster than `mavenCentral()`.

File: src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/comand/CellDecrement.java [hunk 24]
```diff
@@ -1,8 +1,32 @@
...
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
 package com.teamdev.javaclasses.aleksandrov.brainfuck.comand;
 
 import com.teamdev.javaclasses.aleksandrov.brainfuck.compiler.CommandVisitor;
 
-public class CellDecrement implements Command{
+/**
+ * Accepts command visitor
```
Comment (alexander-yevsyukov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/comand/CellDecrement.java]: We need dot at the end of the line.

Comment (alexander-yevsyukov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/comand/CellDecrement.java]: We have an operation here. The fact that it implements Visitor pattern may be described in the parent interface. What we need here is to describe why a new class or interface exists.

Comment (alexander-aleksandrov) [inline src/main/java/com/teamdev/javaclasses/aleksandrov/brainfuck/comand/CellDecrement.java]: Fixed

File: build.gradle (hunk 23 above)
Comment (alexander-aleksandrov) [inline build.gradle]: Fixed

**Pull request merged**